sdk/core/System.ClientModel/src/ModelReaderWriter/ModelWriterOfT.SequenceBuilder.cs

```diff
@@ -0,0 +1,177 @@
+                if (free.Length >= sizeHint)
+                {
+                    return free;
+                }
+
+                // else allocate a new buffer:
+                ExpandBuffers(sizeToRent);
+
+                return _buffers[_count - 1].Array;
+            }
+
+            private readonly object _lock = new object();
+            private void ExpandBuffers(int sizeToRent)
+            {
+                lock (_lock)
+                {
+                    int bufferCount = _count == 0 ? 1 : _count * 2;
+
+                    Buffer[] resized = new Buffer[bufferCount];
+                    if (_count > 0)
+                    {
+                        _buffers.CopyTo(resized, 0);
+                    }
+                    _buffers = resized;
+                    _buffers[_count].Array = ArrayPool<byte>.Shared.Rent(sizeToRent);
+                    _count = bufferCount == 1 ? bufferCount : _count + 1;
+                }
+            }
+
+            /// <summary>
+            /// Returns a <see cref="Span{T}"/> to write to that is at least the requested size, as specified by the <paramref name="sizeHint"/> parameter.
+            /// </summary>
+            /// <param name="sizeHint">The minimum length of the returned <see cref="Span{T}"/>. If less than 256, a buffer of size 256 will be returned.</param>
+            /// <returns>A buffer of at least <paramref name="sizeHint"/> bytes. If <paramref name="sizeHint"/> is less than 256, a buffer of size 256 will be returned.</returns>
+            public Span<byte> GetSpan(int sizeHint = 0)
+            {
+                Memory<byte> memory = GetMemory(sizeHint);
+                return memory.Span;
+            }
+
+            /// <summary>
+            /// Disposes the SequenceWriter and returns the underlying buffers to the pool.
+            /// </summary>
+            public void Dispose()
+            {
+                int bufferCountToFree;
+                Buffer[] buffersToFree;
+                lock (_lock)
+                {
+                    bufferCountToFree = _count;
+                    buffersToFree = _buffers;
+                    _count = 0;
+                    _buffers = Array.Empty<Buffer>();
+                }
+
+                for (int i = 0; i < bufferCountToFree; i++)
+                {
+                    ArrayPool<byte>.Shared.Return(buffersToFree[i].Array);
+                }
+            }
+
+            public bool TryComputeLength(out long length)
+            {
+                length = 0;
+                for (int i = 0; i < _count; i++)
+                {
+                    length += _buffers[i].Written;
+                }
+                return true;
+            }
+
+            public void CopyTo(Stream stream, CancellationToken cancellation)
+            {
+                for (int i = 0; i < _count; i++)
+                {
+                    if (cancellation.IsCancellationRequested)
+                        break;
```

Comment:
Should this throw TaskCancelledException?

Reply:
I didn't really see great recommendations on when to throw and when not to.  Based on this blog https://devblogs.microsoft.com/premier-developer/recommended-patterns-for-cancellationtoken/ best I can gather is maybe if we have finished we should not throw but it doesn't really explicitly say that if you stop in the middle you should throw?

Reply:
we should throw https://learn.microsoft.com/en-us/dotnet/api/system.operationcanceledexception?view=net-7.0